hermes/bin/src/runtime_extensions/hermes/crypto/state.rs

```diff
@@ -0,0 +1,266 @@
+
+/// Map of app name to resource holder
+type State = DashMap<HermesAppName, ResourceHolder>;
+
+/// Wrapper for `XPrv` to implement Hash used in `DashMap`
+#[derive(Eq, Clone, PartialEq)]
+struct WrappedXPrv(XPrv);
+
+/// Implement Hash for `WrappedXPrv`
+impl Hash for WrappedXPrv {
+    /// Hasher for `WrappedXPrv`
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.as_ref().hash(state);
+    }
+}
+
+/// Implement From for `XPrv` to `WrappedXPrv`
+impl From<XPrv> for WrappedXPrv {
+    /// Turn `XPrv` into `WrappedXPrv`
+    fn from(xprv: XPrv) -> Self {
+        WrappedXPrv(xprv)
+    }
+}
+
+/// Resource holder to hold the resources for `XPrv`.
+#[derive(Clone)]
+pub(crate) struct ResourceHolder {
+    /// Map of resource to id.
+    id_to_resource_map: DashMap<u32, XPrv>,
+    /// Map of id to resource.
+    resource_to_id_map: DashMap<WrappedXPrv, u32>,
+    /// Current Id.
+    current_id: u32,
+}
+
+impl ResourceHolder {
+    /// Generate new resource holder.
+    fn new() -> Self {
+        Self {
+            id_to_resource_map: DashMap::new(),
+            resource_to_id_map: DashMap::new(),
+            current_id: 0,
+        }
+    }
+
+    /// Get the next id and increment the current id.
+    fn get_and_increment_next_id(&mut self) -> u32 {
+        self.current_id += 1;
+        self.current_id
+    }
+
+    /// Get the resource from id if possible.
+    fn get_resource_from_id(&self, id: u32) -> Option<XPrv> {
+        self.id_to_resource_map
+            .get(&id)
+            .map(|entry| entry.value().clone())
+    }
+
+    /// Get the id from resource if possible.
+    fn get_id_from_resource(&self, resource: &XPrv) -> Option<u32> {
+        self.resource_to_id_map
+            .get(&WrappedXPrv::from(resource.clone()))
+            .map(|entry| *entry.value())
+    }
+
+    /// Drop the item from resources using id if possible.
+    /// Return the id of the resource if successful remove from maps
+    fn drop(&mut self, id: u32) -> Option<u32> {
+        // Check if the resource exists in id_to_resource_map.
+        if let Some(resource) = self.get_resource_from_id(id) {
+            // Check if the id exists in resource_to_id_map.
+            if let Some(associated_id) = self.get_id_from_resource(&resource) {
+                // The id should be the same.
+                if associated_id == id {
+                    // Remove the resource from both maps.
+                    if let Some(r) = self.id_to_resource_map.remove(&id) {
+                        self.resource_to_id_map.remove(&WrappedXPrv(r.1));
+                        return Some(associated_id);
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+/// Global state to hold the resources.
+static CRYPTO_INTERNAL_STATE: Lazy<Arc<State>> = Lazy::new(|| Arc::new(DashMap::new()));
+
+/// Get the state.
+pub(crate) fn get_state() -> Arc<State> {
+    CRYPTO_INTERNAL_STATE.clone()
+}
```

Comment:
Complementing my comment above, I'm almost certain that this can be changed to `&'static State` since `Lazy<T>` is `static` and implements `Deref<Target = T>` (meaning: it's safe to remove the `Arc`).

PS: Maybe this can be `pub(super)` since it's meant to only be used internally by `mod crypto`.